presto-iceberg/src/main/java/com/facebook/presto/iceberg/TableStatisticsMaker.java

```diff
@@ -199,6 +229,68 @@ private TableStatistics makeTableStatistics(IcebergTableHandle tableHandle, Cons
         return result.build();
     }
 
+    public static void writeTableStatistics(NodeVersion nodeVersion, TypeManager typeManager, IcebergTableHandle tableHandle, Table icebergTable, ConnectorSession session, Collection<ComputedStatistics> computedStatistics)
+    {
+        new TableStatisticsMaker(typeManager, icebergTable).writeTableStatistics(nodeVersion, session, tableHandle, computedStatistics);
+    }
+
+    private void writeTableStatistics(NodeVersion nodeVersion, ConnectorSession session, IcebergTableHandle tableHandle, Collection<ComputedStatistics> computedStatistics)
+    {
+        Snapshot snapshot = tableHandle.getSnapshotId().map(icebergTable::snapshot).orElseGet(icebergTable::currentSnapshot);
+        if (snapshot == null) {
+            // this may occur if the table has not been written to.
+            return;
+        }
+        try (FileIO io = icebergTable.io()) {
+            String path = ((HasTableOperations) icebergTable).operations().metadataFileLocation(format("%s-%s.stats", session.getQueryId(), randomUUID()));
+            OutputFile outputFile = io.newOutputFile(path);
+            try (PuffinWriter writer = Puffin.write(outputFile)
+                    .createdBy("presto-" + nodeVersion)
+                    .build()) {
+                computedStatistics.stream()
+                        .map(ComputedStatistics::getColumnStatistics)
+                        .filter(Objects::nonNull)
+                        .flatMap(map -> map.entrySet().stream())
+                        .collect(Collectors.toMap(Map.Entry::getKey, Map.Entry::getValue))
+                        .forEach((key, value) -> {
+                            if (!key.getStatisticType().equals(NUMBER_OF_DISTINCT_VALUES)) {
+                                return;
+                            }
+                            Optional<Integer> id = Optional.ofNullable(icebergTable.schema().findField(key.getColumnName())).map(Types.NestedField::fieldId);
+                            if (!id.isPresent()) {
+                                log.warn("failed to find column name %s in schema of table %s when writing distinct value statistics", key.getColumnName(), icebergTable.name());
```

Comment:
Returning an error would abort the whole operation and cause the user to have to re-scan the table. It seems like wasted work to throw away the rest of the computed stats just because one column changed IMO.

I could potentially skip any errors, write the available data, but still return an error with the suppressed exceptions after the other files have successfully been written? In that case, the user `ANALYZE` would fail, but some statistics would be written which also seems wrong. I don't have a strong preference for either, it just feels bad to throw away the computed statistics due to one failure

Reply:
The most recent fix now fails-fast when this exception is encountered